### PR TITLE
fix: thorchain utxo unknown fees

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

This PR ensures we get fees estimation back at input time for UTXOs, by leveraging xpub when available. This is after https://github.com/shapeshift/web/pull/8079 went in, which made UTXO fees unknown for UTXOs regardless of wallet connected state, and only got the fees on final quote getting before sign and broadcast. 

While this was a feature, ops rightly pointed out that it effectively looks and feels like a regression, and if it does so, it probably is.
Fixed by leveraging xpub if handy (i.e if wallet connected) to actually estimate things.

```
TODO: 

- [x] Test me with and without a wallet connected 
  - [x] ensure this doesn't rug quotes with UTXO sells without a wallet connected 
  - [x] ensure this doesn't rug execution with a wallet connected
- [x] cleanup in a nice enough way so that us deviating from TradeRateInput types with what's effectively a quoteish input isn't too messy
```

## Issue (if applicable)

closes #

## Risk

> High Risk PRs Require 2 approvals

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

## Screenshots (if applicable)

- No wallet connected

image

- Wallet connected

https://jam.dev/c/873fc60a-82fe-429f-a274-b99f05553e65

### CODE REVIEW

File: packages/swapper/src/swappers/ThorchainSwapper/utils/getL1Rate.ts [hunk 1]
```diff
@@ -306,27 +313,51 @@ export const getL1Rate = async (
...
+              if (!(input as GetUtxoTradeQuoteInput).xpub)
+                return {
+                  networkFeeCryptoBaseUnit: undefined,
+                  protocolFees: getProtocolFees(quote),
+                }
+
+              // This works by leveraging the receiveAddress if it exists. Even though this is a rate, we absolutely can (and will if possible) pass a receive address
+              // However, we run the risk that https://github.com/shapeshift/web/issues/7979 breaks this, since connecting a wallet won't refetch a quote
+              // If that becomes an issue we should be able to get a very rough estimation (not taking users' UTXOs into account) without an address
+              // using sats per byte and byte size from memo. Yes, we don't have a memo returned, but can build it in-house for this purpose easily.
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ThorchainSwapper/utils/getL1Rate.ts:342]: @0xApotheosis FYI ☝🏽

File: packages/swapper/src/swappers/ThorchainSwapper/utils/getL1Rate.ts [hunk 2]
```diff
@@ -306,27 +313,51 @@ export const getL1Rate = async (
...
-            }
+            const feeData = await (async () => {
+              // This is a rate without a wallet connected, so we can't get fees
+              if (!(input as GetUtxoTradeQuoteInput).xpub)
+                return {
+                  networkFeeCryptoBaseUnit: undefined,
+                  protocolFees: getProtocolFees(quote),
+                }
+
+              // This works by leveraging the receiveAddress if it exists. Even though this is a rate, we absolutely can (and will if possible) pass a receive address
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ThorchainSwapper/utils/getL1Rate.ts:339]:
```suggestion
              // This works by leveraging the xpub if it exists. Even though this is a rate, we absolutely can (and will if possible) pass a receive address
```

File: src/components/MultiHopTrade/hooks/useGetTradeQuotes/getTradeQuoteInput.ts [hunk 3]
```diff
@@ -109,7 +109,8 @@ export const getTradeQuoteInput = async ({
     }
 
     case CHAIN_NAMESPACE.Utxo: {
-      if (!(quoteOrRate === 'quote' && wallet))
+      // This is a quote without a wallet, we monkey-patch things to the best of our ability
+      if (quoteOrRate === 'rate' && !receiveAddress)
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/hooks/useGetTradeQuotes/getTradeQuoteInput.ts:113]: Shoudl probably keep on using the wallet discriminator here, we're interested in xpub, not receiveAddress

**Review by 0xApotheosis — APPROVED** [review]
✅ Gas fee shows for UTXOs when wallet is connected
✅ Gas fee still unknown for UTXOs without wallet connected
✅ Connecting a wallet after getting a UTXO quote correctly fetches the gas fees

**Pull request merged**